File: verifier-cli/src/main/java/com/jetbrains/pluginverifier/verifiers/clazz/SuperClassVerifier.java
package com.jetbrains.pluginverifier.verifiers.clazz;

import com.intellij.structure.resolvers.Resolver;
import com.jetbrains.pluginverifier.location.ProblemLocation;
import com.jetbrains.pluginverifier.problems.ClassNotFoundProblem;
import com.jetbrains.pluginverifier.problems.IncompatibleClassChangeProblem;
import com.jetbrains.pluginverifier.verifiers.VerificationContext;
import com.jetbrains.pluginverifier.verifiers.util.VerifierUtil;
import org.objectweb.asm.tree.ClassNode;

/**
 * Check that superclass exists.
 *
 * @author Dennis.Ushakov
 */
public class SuperClassVerifier implements ClassVerifier {
  public void verify(final ClassNode clazz, final Resolver resolver, final VerificationContext ctx) {
    final String superClassName = clazz.superName == null ? "java/lang/Object" : clazz.superName;
    ClassNode aClass = VerifierUtil.findClass(resolver, superClassName, ctx);
    if (aClass == null) {
      if (!ctx.getVerifierOptions().isExternalClass(superClassName)) {
        ctx.registerProblem(new ClassNotFoundProblem(superClassName), ProblemLocation.fromClass(clazz.name));
      }
      return;
    }
    if (VerifierUtil.isInterface(aClass)) {
      ctx.registerProblem(new IncompatibleClassChangeProblem(superClassName, IncompatibleClassChangeProblem.Change.CLASS_TO_INTERFACE), ProblemLocation.fromClass(clazz.name));
    }
  }
}

File: verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/clazz/SuperClassVerifier.kt
package com.jetbrains.pluginverifier.verifiers.clazz

import com.jetbrains.pluginverifier.results.problems.SuperClassBecameInterfaceProblem
import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.createClassLocation
import com.jetbrains.pluginverifier.verifiers.isInterface
import com.jetbrains.pluginverifier.verifiers.logic.CommonClassNames
import com.jetbrains.pluginverifier.verifiers.resolveClassOrProblem
import org.objectweb.asm.tree.ClassNode

/**
 * Check that superclass exists and is indeed a class (not interface).
 *
 * @author Dennis.Ushakov
 */
class SuperClassVerifier : ClassVerifier {
  override fun verify(clazz: ClassNode, ctx: VerificationContext) {
    val superClassName = clazz.superName ?: CommonClassNames.JAVA_LANG_OBJECT
    val superNode = ctx.resolveClassOrProblem(superClassName, clazz) { clazz.createClassLocation() } ?: return
    //If the class or interface named as the direct superclass of C is in fact an interface, loading throws an IncompatibleClassChangeError.
    if (superNode.isInterface()) {
      ctx.registerProblem(SuperClassBecameInterfaceProblem(clazz.createClassLocation(), superNode.createClassLocation()))
    }
  }
}
